tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion_test.cc

```diff
@@ -598,6 +598,46 @@ TEST_F(HorizontalLoopFusionTest, TraversalOrder) {
   EXPECT_EQ(total_fusion_instrs, 2);
 }
 
+TEST_F(HorizontalLoopFusionTest, CopyInsertionFusion) {
+  const char* hlo_text = R"(
+HloModule cluster
+
+ENTRY main {
+  cst = f32[1]{0} constant({0})
+  ROOT tuple_out = (f32[1]{0}, f32[1]{0}, f32[1]{0}, f32[1]{0}) tuple(cst, cst, cst, cst)
+}
+)";
+  EXPECT_TRUE(RunAndCompare(hlo_text, ErrorSpec{0,0}));
+
+  auto module = ParseAndReturnVerifiedModule(hlo_text).ValueOrDie();
+  std::unique_ptr<HloModule> compiled_module =
```

Comment:
Instead of running all HLO passes, would it make more sense to only run the relevant one ? Or do you want to test gpu_compiler running those in particular?

Reply:
I wanted to test the e2e compilation flow as the current issue was and e2e one.
The optimization pass itself was already working. So the test you suggest would pass without this PR. This is why we need an e2e test.

Reply:
Makes sense! But in this case I think it would make more sense to create a new file (maybe call it gpu_compiler_test?) and test the entire pipeline there? horizontal_loop_fusion_test should test only HorizontalLoopFusion imo

Reply:
I can work on that tomorrow. This PR was started 22 days ago, so hopefully no more changes will be needed after this one.

Did you find the issue with the CI? Moving thing to a another file won't fix this issue.

Reply:
Hi Frederic,

Indeed we know it's annoying and we are making progress on making it work better: e.g. the rotation was established just this week, some compiler warnings were unified between internal/external builds, so we'll get less round trips. Please bear with us and help us to figure out a good process!

On this PR in particular:

1. The test runs horizontal loop fusion pass in all subtests, so adding one subtest running the whole pipeline which tests not horizontal_loop_fusion but gpu_compiler.cc is not, strictly speaking, correct. Having said that, I have missed it in the previous review, so it would be fair to merge as is now, and we could refactor it later. The reason I brought it up now is that having this test here causes the failure of `horizontal_loop_fusion_test`, as it does not run with GPU backend by default (unless built with --config=cuda). This should be reproducible upstream, if you build without GPU support and try to run this test. Such behavior is desirable, as we do eventually want to compile for GPUs on machines without GPU support.

2. On how to fix it: there are multiple ways to do so. Simplest would be to check the value of platform at runtime, and skip the test if it's not GPU. Alternatives would include using an `xla_test` BUILD macro instead, with `backends` set to `gpu`, guarding with `if_gpu_is_configured` or (worst) using `#if GOOGLE_CUDA` check.

Reply:
Additionally, adding `tests/gpu_codegen_test` to deps seems to force it to only run on GPU.

Reply:
I moved the test to a new file and used xla_test(). But I'm still able to reproduce the issue.
I tried to do as tests/gpu_codegen_test, but I can still reproduce the issue.
So I went with testing the platform and skipping the test when it isn't CUDA.